### PR TITLE
fix: watch only specs matching specPattern

### PR DESCRIPTION
**Pull request by estrada9166**
- Closes  Second part of https://github.com/cypress-io/cypress/issues/22054#issuecomment-1145214532

### Additional details

- We were watching all the files instead of watching the ones related with the specPattern

### PR Tasks

- [ ] Have tests been added/updated?
- [ ] Has the original issue (or this PR, if no issue exists) been tagged with a release in ZenHub? (user-facing changes only)
- [ ] Has a PR for user-facing changes been opened in [`cypress-documentation`](https://github.com/cypress-io/cypress-documentation)? 
- [ ] Have API changes been updated in the [`type definitions`](https://github.com/cypress-io/cypress/blob/develop/cli/types/cypress.d.ts)?

### CODE REVIEW

**Review by tgriesser — CHANGES_REQUESTED** [review]
I think the changes seems like it'll work, but can we some tests confirming the behavior of the functionality here - in the unit test layer should be fine, but current unit test added doesn't tell us anything more than that there's an extra function added to the `ignored`. 

Maybe something like creating a directory with a few files that we wouldn't expect to be watched and make sure that it's not called?

We could create a `_makeSpecWatcher` method that just returns the `specWatcher` instance in so it's easier to test in isolation:

```
this._specWatcher = this._makeSpecWatcher()
```

**Review by tgriesser — APPROVED** [review]

File: packages/data-context/src/sources/ProjectDataSource.ts [hunk 1]
```diff
@@ -335,17 +336,40 @@ export class ProjectDataSource {
...
+    return chokidar.watch('.', {
+      ignoreInitial: true,
+      ignorePermissionErrors: true,
+      cwd: projectRoot,
+      ignored: ['**/node_modules/**', ...excludeSpecPattern, ...additionalIgnorePattern, (file: string) => {
+        // don't ignore things that look like directories,
+        // using statSync, falling back to whether it has an extension name
+        // if that were to throw
+        try {
+          if (fs.statSync(file).isDirectory()) {
```
Comment (flotwig) [inline packages/data-context/src/sources/ProjectDataSource.ts:360]: `ignored`'s second argument has `Fs.Stats`, so you don't need to do a synchronous `stat` here: https://github.com/paulmillr/chokidar#path-filtering

Also, avoid doing synchronous FS operations, it blocks the entire server when you do this.

File: packages/data-context/test/unit/sources/ProjectDataSource.spec.ts [hunk 2]
```diff
@@ -442,6 +444,74 @@ describe('getPathFromSpecPattern', () => {
...
+
+    const SUPPORT_FILE = path.join(this.specWatcherPath, 'cypress', 'support', 'e2e.js')
+    const SPEC_FILE1 = path.join(this.specWatcherPath, 'cypress', 'e2e', 'foo.cy.js')
+    const SPEC_FILE2 = path.join(this.specWatcherPath, 'cypress', 'e2e', 'some', 'new', 'folder', 'foo.cy.js')
+
+    ctx.actions.file.writeFileInProject(SUPPORT_FILE, '// foo')
+    ctx.actions.file.writeFileInProject(SPEC_FILE1, '// foo')
+    ctx.actions.file.writeFileInProject(SPEC_FILE2, '// foo')
+
+    await delay(500)
```
Comment (flotwig) [inline packages/data-context/test/unit/sources/ProjectDataSource.spec.ts:506]: This is bad practice for unit tests, it makes the tests really slow if we have delays like this. Is there anything you could `await` or use `done` instead?

File: packages/data-context/test/unit/sources/ProjectDataSource.spec.ts [hunk 3]
```diff
@@ -442,6 +444,74 @@ describe('getPathFromSpecPattern', () => {
...
+    expect(Array.from(allFiles)).to.eql([
+      SPEC_FILE1,
+      SPEC_FILE2,
+    // Paths don't start with '/' when emitted
+    ].map((p) => p.slice(1)))
+  })
```
Comment (flotwig) [inline packages/data-context/test/unit/sources/ProjectDataSource.spec.ts:512]: So is the assertion here that we "aren't watching `SUPPORT_FILE`"? I think a comment or an explicit assertion would make it clearer:

```suggestion
    // by the end, we should only be watching specs, not support
    expect(Array.from(allFiles)).to.eql([
      SPEC_FILE1,
      SPEC_FILE2,
    // Paths don't start with '/' when emitted
    ].map((p) => p.slice(1)))
  })
```

File: packages/data-context/src/sources/ProjectDataSource.ts (hunk 1 above)
Comment (tgriesser) [reply packages/data-context/src/sources/ProjectDataSource.ts:360]: Updating

**Review by flotwig — CHANGES_REQUESTED** [review]
Looks good, couple small things

File: packages/data-context/test/unit/sources/ProjectDataSource.spec.ts [hunk 4]
```diff
@@ -467,48 +470,83 @@ describe('_makeSpecWatcher', () => {
...
+    specWatcher.on('add', (filePath) => allFiles.add(filePath))
+    specWatcher.on('change', (filePath) => allFiles.add(filePath))
+
+    writeFiles()
+
+    let attempt = 0
+
+    while (allFiles.size < 3 && attempt++ <= 100) {
+      await delay(10)
+    }
```
Comment (flotwig) [inline packages/data-context/test/unit/sources/ProjectDataSource.spec.ts:540]: eheheheh ok :smile:

**Review by flotwig — APPROVED** [review]

File: packages/data-context/test/unit/sources/ProjectDataSource.spec.ts (hunk 3 above)
Comment (flotwig) [reply packages/data-context/test/unit/sources/ProjectDataSource.spec.ts]: Tests were made more descriptive in another way, resolved

File: packages/data-context/src/sources/ProjectDataSource.ts [hunk 5]
```diff
@@ -11,6 +11,7 @@ import { defaultSpecPattern } from '@packages/config'
 import parseGlob from 'parse-glob'
 import micromatch from 'micromatch'
 import RandExp from 'randexp'
+import fs from 'fs'
```
Comment (lmiller1990) [inline packages/data-context/src/sources/ProjectDataSource.ts:14]: Should this be `fs-extra`? I think we prefer this over `fs` everywhere. We should lint against this, probably.

**Review by lmiller1990 — APPROVED** [review]
Seems fine, I wonder if we eventually want to watch `chokidar` to work around all the weirdness, and possibly include a `gitignore: true` arg like `globby`...

File: packages/data-context/src/sources/ProjectDataSource.ts [hunk 6]
```diff
@@ -335,17 +336,46 @@ export class ProjectDataSource {
...
     // the 'all' event includes: add, addDir, change, unlink, unlinkDir
     this._specWatcher.on('all', onProjectFileSystemChange)
   }
 
+  _makeSpecWatcher ({ projectRoot, specPattern, excludeSpecPattern, additionalIgnorePattern }: { projectRoot: string, excludeSpecPattern: string[], additionalIgnorePattern: string[], specPattern: string[] }) {
+    return chokidar.watch('.', {
+      ignoreInitial: true,
+      ignorePermissionErrors: true,
+      cwd: projectRoot,
+      ignored: ['**/node_modules/**', ...excludeSpecPattern, ...additionalIgnorePattern, (file: string, stats?: fs.Stats) => {
```
Comment (lmiller1990) [inline packages/data-context/src/sources/ProjectDataSource.ts:355]:
```suggestion
    return chokidar.watch(projectRoot, {
      ignoreInitial: true,
      ignorePermissionErrors: true,
      ignored: ['**/node_modules/**', ...excludeSpecPattern, ...additionalIgnorePattern, (file: string, stats?: fs.Stats) => {
```

File: packages/data-context/src/sources/ProjectDataSource.ts [hunk 7]
```diff
@@ -335,17 +336,46 @@ export class ProjectDataSource {
...
+        // We need stats arg to make the determination of whether to watch it, because we need to watch directories
+        // chokidar is extremely inconsistent in whether or not it has the stats arg internally
```
Comment (lmiller1990) [inline packages/data-context/src/sources/ProjectDataSource.ts:357]: I also saw some comments around `**/node_modules/**` not always working as expected: https://github.com/paulmillr/chokidar/issues/773

Could do [like this](https://github.com/paulmillr/chokidar/issues/773#issuecomment-536793469)

```js
if (file.includes('node_modules')) {
  return true
}
```

To be extra safe?

**Review by lmiller1990 — APPROVED** [review]
Seems generally find, added some comments around code style + potential improvement based on some historical comments about chokidar's weirdness around ignoring globs and the fact it doesn't always seem to work.

**Review by lmiller1990 — COMMENTED** [review]
Added some non blocking comments.

**Pull request merged**